### PR TITLE
refactor!: optimize sync behaviour

### PR DESCRIPTION
**Pull request by sillyElvy**
## About

Optimizes sync behaviour to not consider weird discord API returns; makes extensions sync correctly

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [ ] I've made this pull request for/as: (check all that apply)
  - [ ] Documentation
  - [x] Breaking change
  - [ ] New feature/enhancement
  - [x] Bugfix

### CODE REVIEW

File: interactions/client/bot.py [hunk 1]
```diff
@@ -153,7 +154,10 @@ async def __compare_sync(self, data: dict, pool: List[dict]) -> Tuple[bool, dict
         attrs: List[str] = [
             name
             for name in ApplicationCommand.__slots__
-            if not name.startswith("_") and not name.endswith("id") and name != "version"
+            if not name.startswith("_")
+            and not name.endswith("id")
+            and name != "version"
+            and name != "default_permission"
```
Comment (Toricane) [inline interactions/client/bot.py]: Would this be better?
```suggestion
            and name not in {"version", "default_permission"}
```

File: interactions/client/bot.py [hunk 2]
```diff
@@ -271,6 +283,14 @@ async def __compare_sync(self, data: dict, pool: List[dict]) -> Tuple[bool, dict
                             # This is an API/Version difference.
                             continue
 
+                    elif (
+                        attr == "dm_permission"
+                        and data.get(attr) == True  # noqa: E712
+                        and command.get(attr) == None  # noqa: E711
+                    ):
+                        # idk it encountered to me and synced unintentional
```
Comment (Toricane) [inline interactions/client/bot.py]:
```suggestion
                        # idk, it encountered me and synced unintentionally
```

File: interactions/client/bot.py [hunk 3]
```diff
@@ -708,22 +728,27 @@ def __check_options(_option: Option, _names: list, _sub_command: Option = MISSIN
...
-            elif _sub_groups_present and len(coro.__code__.co_varnames) < 3:
+            elif _sub_groups_present and len(coro.__code__.co_varnames) < (4 if _isMethod else 3):
                 raise InteractionException(
                     11,
                     message="Your command needs one argument for the sub_command and one for the sub_command_group.",
                 )
-            add: int = 1 + abs(_sub_cmds_present) + abs(_sub_groups_present)
+            add: int = (
+                1 + abs(_sub_cmds_present) + abs(_sub_groups_present) + 1 if _isMethod else +0
+            )
```
Comment (Toricane) [inline interactions/client/bot.py]: Why camelCase?
Also why `+0`????
```suggestion
            _is_method = hasattr(coro, "__func__")
            if not len(coro.__code__.co_varnames) ^ (
                _is_method and len(coro.__code__.co_varnames) == 1
            ):
                raise InteractionException(
                    11, message="Your command needs at least one argument to return context."
                )
            elif "kwargs" in coro.__code__.co_varnames:
                return
            elif _sub_cmds_present and len(coro.__code__.co_varnames) < (3 if _is_method else 2):
                raise InteractionException(
                    11, message="Your command needs one argument for the sub_command."
                )
            elif _sub_groups_present and len(coro.__code__.co_varnames) < (4 if _is_method else 3):
                raise InteractionException(
                    11,
                    message="Your command needs one argument for the sub_command and one for the sub_command_group.",
                )
            add: int = (
                1 + abs(_sub_cmds_present) + abs(_sub_groups_present) + 1 if _is_method else 0
            )
```

**Review by Toricane — CHANGES_REQUESTED** [review]
A few suggestions
Other than that, it looks good

File: interactions/client/bot.py (hunk 3 above)
Comment (sillyElvy) [reply interactions/client/bot.py]: camel case will be resolved, 
grafik

**Review by Toricane — APPROVED** [review]
LGTM!

**Pull request merged**